stored_file.php

```diff
@@ -0,0 +1,380 @@
+		'csv',   // csv data
+		'txt',   // text file
+		'zip',   // zip file
+		
+	);
+
+	public $id;
+	public $extension;
+	public $mime;
+	public $filename;
+	public $source;
+	
+	protected $__moved;
+	
+	public function actual_filename()
+	{
+		if ($this->__moved)
+			return static::resolve_file($this->filename);
+		return $this->source;
+	}
+	
+	public function has_supported_extension($supported_exts = null)
+	{
+		if (!$this->extension)
+			$this->extension = static::parse_extension($this->filename);
+		
+		if (!$this->extension)
+			return false;
+		
+		if ($supported_exts && is_array($supported_exts) && count($supported_exts))
+		{
+			return in_array($this->extension,
+			$supported_exts);
+		}
+
+		return in_array($this->extension,
+			static::$__supported_exts);
+	}
+	
+	public function exists()
+	{
+		return is_file($this->actual_filename());
+	}
+	
+	public function size()
+	{
+		return filesize($this->actual_filename());
+	}
+
+	public function human_size()
+	{
+		$size = $this->size();
+		
+		if ($size > static::SIZE_GIB)
+		{
+			$size = ($size / static::SIZE_GIB);
+			return sprintf('%.2f GiB', $size);
+		}
+		
+		if ($size > static::SIZE_MIB)
+		{
+			$size = ($size / static::SIZE_MIB);
+			return sprintf('%.2f MiB', $size);
+		}
+		
+		$size = ($size / static::SIZE_KIB);
+		return sprintf('%.2f KiB', $size);
+	}
+	
+	public function detect_mime()
+	{
+		if ($this->mime) return $this->mime;
+		$this->mime = File_Util::detect_mime($this->actual_filename());
+		return $this->mime;
+	}
+	
+	public function generate_filename($ext = null)
+	{
+		if (!$ext) $ext = $this->extension;
+		
+		$hash = md5_file($this->source);
+		$dir1 = substr($hash, 0, 2);
+		$dir2 = substr($hash, 2, 2);
+		$name = substr($hash, 4, 28);
+		
+		$this->filename = build_path($dir1, $dir2,
+			sf('%s.%s', $name, $ext));
+		return $this->filename;
+	}
+	
+	public function read()
+	{
+		return file_get_contents($this->actual_filename());
+	}
+	
+	public function move()
+	{
+		if ($this->__moved) return;
+		$destination = static::resolve_file($this->filename, true);
+		$dir = dirname($destination);
+		if (!is_dir($dir)) mkdir($dir, 0755, true);
+		rename($this->source, $destination);
+		$this->__moved = true;
+
+		// File Tracking Feature Scott U 9/2019
+		// Extracting Path and Hash from file destination
+		$stringParts = explode("/", $destination);
+		$hash = $stringParts[6].$stringParts[7].substr($stringParts[8], 0, (strrpos($stringParts[8], ".")));
+		$path = '/'.$stringParts[1].'/'.$stringParts[2].'/'.$stringParts[3].'/'.$stringParts[4].'/'.$stringParts[5].'/'.$stringParts[6].'/'.$stringParts[7].'/';
+
+		// Setup tracking text file location and temp swap file
+		$trackFile = $path.'tracking.txt';
+		$trackFileTemp = $path.'tracking.temp';
+
+		// Check for existing tracking file
+		if (!file_exists($trackFile)) 
+		{   
+			// tracking text file does NOT exist, create it.
+			// Include initial file uploaded count
+			$content = $hash." = 1";
```

Comment:
It seems strange to create a custom file format for this especially when there are lots of suitable formats already available that are well tested and reliable. If someone were to manually edit one of your tracking files changing ` = ` to `=` then it would break everything, despite looking like a reasonable change to the person making it. 

You could have used JSON with `json_encode`/`json_decode`. 
You could have used PHP `serialize()/unserialize()`.
You could have used CSV files. 
You could have had 1 tracking file per stored file with JUST the count inside, reducing complexity greatly. 
... etc ...

Reply:
Minor detail: Inconsistent quotes. You keep switching between single quotes and double quotes for no reason. It's rare that it matters, but sometimes it can cause frustration when searching through (e.g. grep) the entire codebase.